prov/opx/include/rdma/opx/fi_opx_endpoint.h

```diff
@@ -0,0 +1,3167 @@
+	ret = fi_opx_ep_tx_check(opx_ep->tx, av_type);
+	if (ret) return ret;
+#endif
+	assert(dest_addr != FI_ADDR_UNSPEC);
+
+	const union fi_opx_addr addr = {
+		.fi = (av_type == FI_AV_TABLE) ?	/* constant compile-time expression */
+			opx_ep->tx->av_addr[dest_addr].fi :
+			dest_addr
+	};
+
+	const ssize_t rc = FI_OPX_FABRIC_TX_INJECT(ep, buf, len, addr.fi, tag, data,
+			lock_required, addr.hfi1_rx, caps, reliability);
+
+	if (OFI_UNLIKELY(rc == -EAGAIN)) {
+		// In this case we are probably out of replay buffers. To deal
+		// with this, we do a poll which may send a ping and will
+		// process any incoming ACKs, hopefully releasing a buffer for
+		// reuse.
+		fi_opx_ep_rx_poll(&opx_ep->ep_fid, 0, OPX_RELIABILITY,
+			FI_OPX_HDRQ_MASK_RUNTIME);
+	}    
+
+	FI_DBG_TRACE(fi_opx_global.prov, FI_LOG_EP_DATA,
+		"===================================== INJECT (end)\n");
+
+	return rc;
+}
+
+__OPX_FORCE_INLINE__
+ssize_t fi_opx_ep_tx_inject(struct fid_ep *ep,
+			    const void *buf,
+			    size_t len,
+			    fi_addr_t dest_addr,
+			    uint64_t tag,
+			    const uint32_t data,
+			    const int lock_required,
+			    const enum fi_av_type av_type,
+			    const uint64_t caps,
+			    const enum ofi_reliability_kind reliability)
+{
+	struct fi_opx_ep *opx_ep = container_of(ep, struct fi_opx_ep, ep_fid);
+
+	fi_opx_lock_if_required(&opx_ep->lock, lock_required);
+
+	ssize_t rc = fi_opx_ep_tx_inject_internal(ep, buf, len, dest_addr, tag,
+						  data, FI_OPX_LOCK_NOT_REQUIRED, av_type,
+						  caps, reliability);
+
+	fi_opx_unlock_if_required(&opx_ep->lock, lock_required);
+
+	return rc;
+}
+
+__OPX_FORCE_INLINE_AND_FLATTEN__
+ssize_t fi_opx_recv_generic(struct fid_ep *ep,
+			    void *buf, size_t len, void *desc,
+			    fi_addr_t src_addr, uint64_t tag, uint64_t ignore, void *context,
+			    const int lock_required, const enum fi_av_type av_type,
+			    const uint64_t static_flags,
+			    const enum ofi_reliability_kind reliability)
+{
+	struct fi_opx_ep *opx_ep = container_of(ep, struct fi_opx_ep, ep_fid);
+
+	fi_opx_lock_if_required(&opx_ep->lock, lock_required);
+	ssize_t rc = fi_opx_ep_rx_recv_internal(opx_ep, buf, len, desc, src_addr, tag,
+						ignore, context, FI_OPX_LOCK_NOT_REQUIRED, av_type,
+						static_flags, reliability);
+	fi_opx_unlock_if_required(&opx_ep->lock, lock_required);
+
+	return rc;
+}
+
+__OPX_FORCE_INLINE_AND_FLATTEN__
+ssize_t fi_opx_recvmsg_generic(struct fid_ep *ep,
+			       const struct fi_msg *msg, uint64_t flags,
+			       const int lock_required, const enum fi_av_type av_type,
+			       const enum ofi_reliability_kind reliability)
+{
+	struct fi_opx_ep *opx_ep = container_of(ep, struct fi_opx_ep, ep_fid);
+
+	fi_opx_lock_if_required(&opx_ep->lock, lock_required);
+	ssize_t rc = fi_opx_ep_rx_recvmsg_internal(opx_ep, msg, flags, FI_OPX_LOCK_NOT_REQUIRED, av_type, reliability);
+	fi_opx_unlock_if_required(&opx_ep->lock, lock_required);
+
+	return rc;
+}
+
+
+#endif /* _FI_PROV_OPX_ENDPOINT_H_ */
```

Comment:
There are some massively long functions in this file that are forced to be inline.  Do you really find that skipping a function call outweighs the significant increase in code space and the result that has on the instruction cache?